### PR TITLE
Fix Inferencer

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
# Comments

- I have refactored the code a bit and moved the inferencer to `deploy` folder. This is intended to comply with the upcomming refactor
- The place in base `Inferencer` where the meta data is retrieved from the model might require refactor.
- Also, I have changed the default category for `padim` and `stfpm` to bottle as that's the example used in the README.

# Changes

🐞 fix torch inferencer
🛠  Add metadata to inferencer
🧪 Add Inferencer tests

# Tests
image

### CODE REVIEW

File: anomalib/deploy/inference.py [hunk 1]
```diff
@@ -105,6 +108,63 @@ def __call__(self, image: np.ndarray) -> np.ndarray:
...
+        meta_data: Union[Dict, DictConfig],
+    ) -> Tuple[Union[np.ndarray, Tensor], float]:
+        """Applies normalization and resizes the image.
+
+        Args:
+            anomaly_maps (Union[Tensor, np.ndarray]): Predicted raw anomaly map.
+            pred_scores (Union[Tensor, np.float32]): Predicted anomaly score
+            meta_data (Dict): Meta data. Post-processing step sometimes requires
+                additional meta data such as image shape. This variable comprises such info.
+                Defaults to {}.
```
Comment (samet-akcay) [inline anomalib/deploy/inference.py:124]: the default value is not defined above?

File: anomalib/deploy/inference.py [hunk 2]
```diff
@@ -105,6 +108,63 @@ def __call__(self, image: np.ndarray) -> np.ndarray:
...
+            )
+            anomaly_maps = normalize_cdf(anomaly_maps, meta_data["pixel_threshold"])
+
+        # standardize image scores
+        if "image_mean" in meta_data.keys() and "image_std" in meta_data.keys():
+            pred_scores = standardize(pred_scores, meta_data["image_mean"], meta_data["image_std"])
+            pred_scores = normalize_cdf(pred_scores, meta_data["image_threshold"])
+
+        if isinstance(anomaly_maps, Tensor):
+            anomaly_maps = anomaly_maps.numpy()
```
Comment (samet-akcay) [inline anomalib/deploy/inference.py:155]: are we certain that the tensor is on cpu? Perhaps it would be safer to do `anomaly_maps.cpu().numpy()`

File: anomalib/deploy/inference.py [hunk 3]
```diff
@@ -105,6 +108,63 @@ def __call__(self, image: np.ndarray) -> np.ndarray:
...
+
+        if "image_shape" in meta_data and anomaly_maps.shape != meta_data["image_shape"]:
+            anomaly_maps = cv2.resize(anomaly_maps, meta_data["image_shape"])
+
+        return anomaly_maps, float(pred_scores)
+
+    def _load_meta_data(self, path: Optional[Union[str, Path]] = None) -> Union[DictConfig, Dict]:
+        meta_data = {}
```
Comment (samet-akcay) [inline anomalib/deploy/inference.py:163]: How does this pass the tox pipeline without the docstring? Because of _ prefix?

File: anomalib/deploy/inference.py [hunk 4]
```diff
@@ -113,15 +173,57 @@ class TorchInferencer(Inferencer):
...
         path (Union[str, Path]): Path to the model ckpt file.
+        meta_data_path (Union[str, Path], optional): Path to metadata file. If none, it tries to load the params
+                from the model state_dict. Defaults to None.
     """
 
-    def __init__(self, config: Union[DictConfig, ListConfig], path: Union[str, Path, AnomalyModule]):
+    def __init__(
+        self,
+        config: Union[DictConfig, ListConfig],
+        path: Union[str, Path, AnomalyModule],
```
Comment (samet-akcay) [inline anomalib/deploy/inference.py:183]: not sure if `path` is the best name here if it could be `AnomalyModule`

File: tools/inference.py [hunk 5]
```diff
@@ -24,11 +24,7 @@
 import cv2
 
 from anomalib.config import get_configurable_parameters
-from anomalib.core.model.inference import (
-    Inferencer,
-    OpenVINOInferencer,
-    TorchInferencer,
-)
+from anomalib.deploy.inference import Inferencer, OpenVINOInferencer, TorchInferencer
```
Comment (samet-akcay) [inline tools/inference.py:27]: As per the suggestion in #56, should we import openvino only if the weights are ".onnx", ".bin", ".xml"? This means we would need to import not at the top of the module, but we could perhaps ignore such errors.

File: tools/inference.py [hunk 6]
```diff
@@ -42,6 +38,7 @@ def get_args() -> Namespace:
     parser.add_argument("--weight_path", type=str, required=True, help="Path to a model weights")
     parser.add_argument("--image_path", type=str, required=True, help="Path to an image to infer.")
     parser.add_argument("--save_path", type=str, required=False, help="Path to save the output image.")
+    parser.add_argument("--meta_data", type=str, required=False, help="Path to JSON file containing the metadata.")
```
Comment (samet-akcay) [inline tools/inference.py:41]: should we use type `Path` here to automatically convert everything to pathlib.Path object to make it easier to work with the path?

File: tools/inference.py [hunk 7]
```diff
@@ -60,10 +57,10 @@ def infer() -> None:
     extension = os.path.splitext(args.weight_path)[-1]
     inference: Inferencer
     if extension in (".ckpt"):
-        inference = TorchInferencer(config=config, path=args.weight_path)
```
Comment (samet-akcay) [inline tools/inference.py:63]: following the above argument, maybe we could do `from anomalib.deploy.inference TorchInferencer` here?

File: tools/inference.py [hunk 8]
```diff
@@ -60,10 +57,10 @@ def infer() -> None:
...
+        inference = TorchInferencer(config=config, path=args.weight_path, meta_data_path=args.meta_data)
 
     elif extension in (".onnx", ".bin", ".xml"):
-        inference = OpenVINOInferencer(config=config, path=args.weight_path)
```
Comment (samet-akcay) [inline tools/inference.py:66]: same as above, `from anomalib.deploy.inference import OpenVINOInferencer`?

File: anomalib/deploy/inference.py [hunk 9]
```diff
@@ -80,19 +80,22 @@ def predict(
...
         processed_image = self.pre_process(image)
         predictions = self.forward(processed_image)
-        anomaly_map, pred_score = self.post_process(predictions, meta_data=meta_data)
+        anomaly_map, _ = self.post_process(predictions, meta_data=meta_data)
 
         if superimpose is True:
             anomaly_map = superimpose_anomaly_map(anomaly_map, image)
 
-        return anomaly_map, pred_score
+        return anomaly_map
```
Comment (djdameln) [inline anomalib/deploy/inference.py]: I guess this would cause some issues for classification-type algorithms that produce image-level anomaly scores.

Comment (djdameln) [inline anomalib/deploy/inference.py]: In OTE, the OpenVINO task expects both the anomaly map and the predicted score to be returned. We'll need to make some changes there to make it compatible with this design.

File: anomalib/deploy/inference.py [hunk 10]
```diff
@@ -105,6 +108,63 @@ def __call__(self, image: np.ndarray) -> np.ndarray:
...
+        # standardize image scores
+        if "image_mean" in meta_data.keys() and "image_std" in meta_data.keys():
+            pred_scores = standardize(pred_scores, meta_data["image_mean"], meta_data["image_std"])
+            pred_scores = normalize_cdf(pred_scores, meta_data["image_threshold"])
+
+        if isinstance(anomaly_maps, Tensor):
+            anomaly_maps = anomaly_maps.numpy()
+
+        if "image_shape" in meta_data and anomaly_maps.shape != meta_data["image_shape"]:
+            anomaly_maps = cv2.resize(anomaly_maps, meta_data["image_shape"])
```
Comment (djdameln) [inline anomalib/deploy/inference.py]: I feel that the Numpy conversion and resizing should be done outside of the _normalize function, because these operations are not related to the normalization.

File: anomalib/deploy/inference.py [hunk 11]
```diff
@@ -21,7 +21,7 @@
 import cv2
```
Comment (djdameln) [inline anomalib/deploy/inference.py]: Can we rename this file to `inferencer.py`, to better make the distinction between `tools/inference.py`?

File: anomalib/deploy/inference.py [hunk 12]
```diff
@@ -113,15 +173,57 @@ class TorchInferencer(Inferencer):
...
+        Args:
+            path (Optional[Union[str, Path]], optional): Path to metadata file. If none, it tries to load the params
+                from the model state_dict. Defaults to None.
+
+        Returns:
+            Dict: Dictionary containing the meta_data.
+        """
+        meta_data = {}
+        model_params = self.model.state_dict()
+        meta_data_params = {
```
Comment (djdameln) [inline anomalib/deploy/inference.py]: Since the model is already loaded here, you could just get the metadata from the model itself:
```
image_threshold = self.model.image_threshold
pixel_threshold = self.model.pixel_threshold
image_mean = self.model.training_distribution.image_mean
...
```

File: anomalib/deploy/inferencers/base.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+                meta_data = getattr(self, "meta_data")
+            else:
+                meta_data = {}
+        if isinstance(image, (str, Path)):
+            image = read_image(image)
+        meta_data["image_shape"] = image.shape[:2]
+
+        processed_image = self.pre_process(image)
+        predictions = self.forward(processed_image)
+        anomaly_map, _ = self.post_process(predictions, meta_data=meta_data)
```
Comment (djdameln) [inline anomalib/deploy/inferencers/base.py:88]: I'm guessing this would cause some issues for classification-type algorithms. Maybe we can think of a design that supports both segmentation and classification algos.

File: anomalib/deploy/inferencers/base.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+        if superimpose is True:
+            anomaly_map = superimpose_anomaly_map(anomaly_map, image)
+
+        return anomaly_map
```
Comment (djdameln) [inline anomalib/deploy/inferencers/base.py:93]: The OTE OpenVINO task expects that both the anomaly map and the predicted score are returned here. So we'll need to make some changes there to accommodate this design.

File: anomalib/deploy/inferencers/base.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        # standardize image scores
+        if "image_mean" in meta_data.keys() and "image_std" in meta_data.keys():
+            pred_scores = standardize(pred_scores, meta_data["image_mean"], meta_data["image_std"])
+            pred_scores = normalize_cdf(pred_scores, meta_data["image_threshold"])
+
+        if isinstance(anomaly_maps, Tensor):
+            anomaly_maps = anomaly_maps.cpu().numpy()
+
+        if "image_shape" in meta_data and anomaly_maps.shape != meta_data["image_shape"]:
+            anomaly_maps = cv2.resize(anomaly_maps, meta_data["image_shape"])
```
Comment (djdameln) [inline anomalib/deploy/inferencers/base.py:152]: I feel that the Numpy conversion and resizing should be taken outside of the normalize method, because both these operations are not related to normalization.

File: anomalib/deploy/inferencers/torch.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        Args:
+            path (Optional[Union[str, Path]], optional): Path to metadata file. If none, it tries to load the params
+                from the model state_dict. Defaults to None.
+
+        Returns:
+            Dict: Dictionary containing the meta_data.
+        """
+        meta_data: Union[DictConfig, Dict[str, Union[float, np.ndarray, Tensor]]] = {}
+        model_params = self.model.state_dict()
+        meta_data_params = {
```
Comment (djdameln) [inline anomalib/deploy/inferencers/torch.py:69]: The model is already loaded here, so we can just get the metadata from the model itself:
```
image_threshold = self.model.image_threshold.value
pixel_threshold = self.model.pixel_threshold.value
image_mean = self.model.training_distribution.image_mean
...
```

File: anomalib/deploy/inferencers/base.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+
+        return anomaly_maps, float(pred_scores)
+
+    def _load_meta_data(
```
Comment (djdameln) [inline anomalib/deploy/inferencers/base.py:156]: It's not fully clear to me how and where these metadata json files are generated

File: anomalib/core/callbacks/compress.py [hunk 18]
```diff
@@ -30,14 +32,10 @@ def on_train_end(self, trainer, pl_module: LightningModule) -> None:  # pylint:
...
         onnx_path = os.path.join(self.dirpath, self.filename + ".onnx")
-        height, width = self.input_size
-        torch.onnx.export(
-            pl_module.model,
-            torch.zeros((1, 3, height, width)).to(pl_module.device),
-            onnx_path,
-            opset_version=11,
-            input_names=["input"],
-            output_names=["output"],
+        pl_module = cast(AnomalyModule, pl_module)
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/core/callbacks/compress.py:35]: In some of the callback methods we define `pl_module: AnomalyModule`, while in some others `pl_module: LightningModule`. If pylint and mypy are happy with `pl_module: AnomalyModule`, then we could maybe update the others as well. Therefore, we wouldn't need to cast it later in the code.

[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/core/callbacks/compress.py:35]: But I guess this is a topic for another PR

File: anomalib/deploy/inferencers/base.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+        """Pre-process."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def forward(self, image: Union[np.ndarray, Tensor]) -> Union[np.ndarray, Tensor]:
+        """Forward-Pass input to model."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def post_process(self, predictions: Union[np.ndarray, Tensor], meta_data: Optional[Dict]) -> np.ndarray:
```
Comment (samet-akcay) [inline anomalib/deploy/inferencers/base.py:55]: base inference task and torch inference tasks return only np.ndarray, while openvino inferencer returns Tuple[np.ndarray, float]. Design-wise, this is not correct, right? The signature should be the same. Is there a reason why only `np.ndarray` is returned here for anomaly_map, and not scores?

File: tools/inference.py [hunk 20]
```diff
@@ -60,10 +59,14 @@ def infer() -> None:
     extension = os.path.splitext(args.weight_path)[-1]
```
Comment (samet-akcay) [inline tools/inference.py:59]: Now that the paths are of type `Path`, we could maybe use `extension = args.weight_path.suffix` and remove `import os` from the import at the top of the module.

File: anomalib/deploy/inferencers/base.py (hunk 17 above)
Comment (ashwinvaidya17) [reply anomalib/deploy/inferencers/base.py:156]: We don't have metadata json files but this is just to add support for when we add them in the future. We will need this for the OpenVINO inferencer as the model does not contain the metadata.

Comment (djdameln) [reply anomalib/deploy/inferencers/base.py:156]: alright. So currently OpenVINO inference does not support normalization

Comment (ashwinvaidya17) [reply anomalib/deploy/inferencers/base.py:156]: Yes

**Review by samet-akcay — APPROVED** [review]

File: anomalib/core/callbacks/compress.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/core/callbacks/compress.py:35]: Created Issue https://github.com/openvinotoolkit/anomalib/issues/65
